ghcjs-src/Miso/Subscription/WebSocket.hs

```diff
@@ -151,48 +129,207 @@ data SocketState
+defaultPingConfig :: PingConfig
+defaultPingConfig = PingConfig 30 "ping" "pong"
+
+-- | Default websocket config.
+-- Doesn't set up a ping, and attempts to reconnect if close was abnormal.
+defaultWSConfig :: WebSocketConfig
+defaultWSConfig = WebSocketConfig {
+  websocketPing = Nothing,
+  websocketShouldReconnectOnClose = \event -> pure (not $ closedCleanly event)
+  }
+
+--------------------------------------------------------------------------------
+-- * Creating Subscriptions
+
+
+-- | WebSocket subscription with default configuration.
+websocketSub
+  :: FromJSON message
+  => WebSocket                          -- ^ Handle to the socket.
+  -> (WebSocketEvent message -> action) -- ^ Message handler.
+  -> Sub action model                   -- ^ A subscription.
+websocketSub = websocketSubWithConfig defaultWSConfig
+
+
+-- | WebSocket subscription, given a particular configuration.
+--
+-- If the socket disconnects, the config will be used to determine
+-- whether to reconnect; if so it happens seamlessly. If not, the
+-- WebSocketClosed event will be fired.
+websocketSubWithConfig
+  :: FromJSON message
+  => WebSocketConfig                    -- ^ Configuration
+  -> WebSocket                          -- ^ Handle to the socket.
+  -> (WebSocketEvent message -> action) -- ^ Event handler.
+  -> Sub action model                   -- ^ A subscription
+websocketSubWithConfig config socket handler getModel sink = do
+  let (url, protocols) = (wsURL socket, wsProtocols socket)
+
+  -- Spin off ping thread if configured
+  pingThreadId <- case websocketPing config of
+    Nothing -> pure Nothing
+    Just (PingConfig{..}) -> fmap Just $ forkIO $ forever $ do
+      threadDelay (pingSeconds * 1000000)
+      sendTextToWebSocket socket pingSend
+
+  -- Fire the WebSocketOpen event when the socket opens.
+  socket_ <- getSocket_ socket
+  onOpen socket_ =<< do
+    asyncCallback (sink (handler WebSocketOpen))
+
+  -- When a message is received, parse it as JSON and fire an event.
+  onMessage socket_ =<< do
+    let parseData raw = do
+          result <- safeParseString raw
+          sink $ case result of
+            Error err -> handler (WebSocketUnparseableMessage raw (ms err))
+            Success message -> handler (WebSocketMessage message)
+    asyncCallback1 $ case websocketPing config of
+      -- Iff a ping is configured, filter out ping responses before parsing.
+      Just (PingConfig {..}) -> getData >=> \case
+        msg | msg == pingReceive -> pure ()
+        rawData -> parseData rawData
+      Nothing -> getData >=> parseData
+
+  -- When the socket closed, determine whether to reconnect. If so,
+  -- update the MVar in the WebSocket. If reconnection fails, the
+  -- WebSocketConnectFailed event will be fired.
+  onClose socket_ =<< do
+    asyncCallback1 $ \e -> do
+      -- Stop the ping thread, if it's running.
+      maybe (pure ()) killThread pingThreadId
```

Comment:
`traverse_ killThread pingThreadId`

Reply:
works for me

Reply:
Actually I'd kind of prefer using `mapM_` since it's part of the prelude